File: Arrays/StriverSheet/sortColors.js
// https://leetcode.com/problems/sort-colors/

function sortColors(nums) {
    if(nums.length <= 1) return nums;

    let low = 0, mid = 0, high = nums.length - 1;

    while(mid <= high) {
        const num = nums[mid];
        switch(num) {
            case 0: 
                swap(num, low++, mid++);
                break;
            case 1:
                mid++;
                break;
            case 2:
                swap(num, mid, high--);
        }
    }
    return nums;
}

function swap(nums, i, j) {
    [ nums[i], nums[j] ] = [ nums[j], nums[i] ];
}

File: Leetcode/SortColors.java
/*
----- Problem explanation -----
Given an array nums with n objects colored red, white, or blue, sort them in-place so that objects of the same color are adjacent, with the colors in the order red, white, and blue.

We will use the integers 0, 1, and 2 to represent the color red, white, and blue, respectively.

You must solve this problem without using the library's sort function.

Example 1:

Input: nums = [2,0,2,1,1,0]
Output: [0,0,1,1,2,2]

Example 2:

Input: nums = [2,0,1]
Output: [0,1,2]
 
Constraints:

n == nums.length
1 <= n <= 300
nums[i] is either 0, 1, or 2.

----- Approach -----

Take 3 Pointers low = 0, mid = 0 and high = nums.length -1;
Use loop to iterate the array with condition mid <= high.(Since we only need to check middle elements of low and high).
if element is 0 swap with low and low++, mid++.
if element is 1 then mid++.
if element is 2 then swap with high and high--.
*/
class Solution {
    public void sortColors(int[] nums) {
        int low = 0, mid = 0, high = nums.length-1;

        while(mid <= high){
            if(nums[mid] == 0 ){
                //swap with left
                swap(nums,low, mid );
                low++;
                mid++;
            }else if(nums[mid] == 2){
                swap(nums, mid, high);
                high--;
            }else{
                mid++;
            }
        }
        System.out.println(nums);
    }

    public static void swap(int[] nums, int i, int j){
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}
